client/allocrunner/taskrunner/artifact_hook.go

```diff
@@ -60,14 +56,57 @@ func (h *artifactHook) Prestart(ctx context.Context, req *interfaces.TaskPrestar
 			)
 			herr := NewHookError(wrapped, structs.NewTaskEvent(structs.TaskArtifactDownloadFailed).SetDownloadError(wrapped))
 
-			return herr
+			errorChannel <- herr
+			continue
 		}
 
 		// Mark artifact as downloaded to avoid re-downloading due to
 		// retries caused by subsequent artifacts failing. Any
 		// non-empty value works.
 		resp.State[aid] = "1"
 	}
+}
+
+func (*artifactHook) Name() string {
+	// Copied in client/state when upgrading from <0.9 schemas, so if you
+	// change it here you also must change it there.
+	return "artifacts"
+}
+
+func (h *artifactHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if len(req.Task.Artifacts) == 0 {
+		resp.Done = true
+		return nil
+	}
+
+	// Initialize hook state to store download progress
+	resp.State = make(map[string]string, len(req.Task.Artifacts))
+
+	h.eventEmitter.EmitEvent(structs.NewTaskEvent(structs.TaskDownloadingArtifacts))
+
+	// maxConcurrency denotes the number of workers that will download artifacts in parallel
+	maxConcurrency := 3
+
+	// jobsChannel is a buffered channel which will have all the artifacts that needs to be processed
+	jobsChannel := make(chan *structs.TaskArtifact, maxConcurrency)
+	// Push all artifact requests to job channel
+	go func() {
+		for _, artifact := range req.Task.Artifacts {
+			jobsChannel <- artifact
```

Comment:
This needs to select on `errorChannel` otherwise it will block forever and leak a goroutine if `len(req.Task.Artifacts) > maxConcurrency`

Reply:
thanks. resolved now.